### PR TITLE
Introduce withRefreshAuth

### PR DESCRIPTION
**Pull request by djhi**

### CODE REVIEW

**Comment by fzaninotto** [comment]
I don't like the idea of a single decorator for two very different objects. I also see that it forces you to do some borderline inspection to determine in which case you are. How about one for each?

**Comment by djhi** [comment]
> How about one for each?

How would you name them? 
`addRefreshAuthToDataProvider` and `addRefreshAuthToAuthProvider`?

**Comment by fzaninotto** [comment]
yes, fine by me

**Review by slax57 — CHANGES_REQUESTED** [review]

File: docs/addRefreshAuthToDataProvider.md [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+# `addRefreshAuthToDataProvider`
+
+This helper function wraps an existing [`dataProvider`](./DataProviderIntroduction.md) to support authentication token refreshing mechanisms.
+
+## Usage
+
+Use `addRefreshAuthToDataProvider` to decorate an existing data provider. In addition to the base provider, this function takes a function responsible for refreshing the authentication token if needed.
+
+Here is a simple example that refreshes an expired JWT token when needed:
```
Comment (slax57) [inline docs/addRefreshAuthToDataProvider.md:14]: I wonder if we should add this simple example as its own dedicated paragraph in the Security > Introduction page too. What do you think @fzaninotto ?

**Review by slax57 — APPROVED** [review]

File: docs/addRefreshAuthToDataProvider.md [hunk 2]
```diff
@@ -28,52 +28,43 @@ export const refreshAuth = () => {
...
-const myAuthProvider = {
-    // ...Usual AuthProvider methods
-};
-
-export const authProvider = withRefreshAuth(myAuthProvider, refreshAuth);
+export const dataProvider = addRefreshAuthToDataProvider(baseDataProvider, refreshAuth);
 ```
 
-Then, inject the decorated providers in the `<Admin>` component:
+Then, inject the decorated provider in the `<Admin>` component:
```
Comment (djhi) [inline docs/addRefreshAuthToDataProvider.md:40]:
```suggestion
Then, pass the decorated provider to the `<Admin>` component
```

File: docs/addRefreshAuthToAuthProvider.md [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+import { refreshAuth } from 'refreshAuth';
+
+const myAuthProvider = {
+    // ...Usual AuthProvider methods
+};
+
+export const authProvider = addRefreshAuthToAuthProvider(myAuthProvider, refreshAuth);
+```
+
+Then, inject the decorated provider in the `<Admin>` component:
```
Comment (djhi) [inline docs/addRefreshAuthToAuthProvider.md:41]:
```suggestion
Then, pass the decorated provider to the `<Admin>` component
```

File: docs/addRefreshAuthToDataProvider.md (hunk 1 above)
Comment (fzaninotto) [reply docs/addRefreshAuthToDataProvider.md:14]: 👍

Comment (djhi) [reply docs/addRefreshAuthToDataProvider.md:14]: Done

File: docs/Authentication.md [hunk 4]
```diff
@@ -232,6 +232,42 @@ It's up to you to decide when to redirect users to the third party authenticatio
 * Directly in the `AuthProvider.checkAuth()` method as above;
 * When users click a button on a [custom login page](#customizing-the-login-component)
 
+## Handling Refresh Tokens
+
+[Refresh tokens](https://oauth.net/2/refresh-tokens/) are an important security mechanism. In order to leverage them, you should decorate the `dataProvider` and the `authProvider` so that they can check whether the authentication must be refreshed and actually refresh it.
+
+You can use the [`addRefreshAuthToDataProvider`](addRefreshAuthToDataProvider.md) and [`addRefreshAuthToAuthProvider`](./addRefreshAuthToAuthProvider.md) functions for this purpose. They accept a `dataProvider` or `authProvider` respectively and a function that should refresh the authentication token if necessary:
```
Comment (slax57) [inline docs/Authentication.md:239]:
```suggestion
You can use the [`addRefreshAuthToDataProvider`](./addRefreshAuthToDataProvider.md) and [`addRefreshAuthToAuthProvider`](./addRefreshAuthToAuthProvider.md) functions for this purpose. They accept a `dataProvider` or `authProvider` respectively and a function that should refresh the authentication token if necessary:
```

**Review by slax57 — CHANGES_REQUESTED** [review]

**Review by slax57 — APPROVED** [review]

**Comment by djhi** [comment]
Indeed, we could do a better job on the documentation side. One way to avoid the double refresh would be to store the first promise and return it on subsequent calls until it is resolved. You can check how we did it in `ra-directus` for instance: https://github.com/marmelab/ra-directus/blob/main/packages/ra-directus/src/directusRefreshAuthToken.ts

Maybe we should handle this ourselves so you don't have to worry about it though

**Comment by fzaninotto** [comment]
I'v just changed the repo visibility, you should be able to see the code.

**Comment by davidhenley** [comment]
Thanks @fzaninotto! I see it now

**Comment by djhi** [comment]
```tsx
let refreshPromise = null;

const refreshToken = () => {

  if (refreshPromise != null) {
    return refreshPromise;
  }

  const accessToken = localStorage.getItem('authentication.access_token');
  const refreshToken = localStorage.getItem('authentication.refresh_token');

  if (!accessToken && !refreshToken) {
    return Promise.reject();
  }

  const now = new Date()
  const expired = jwtDecode(accessToken).exp < now.getTime() / 1000

  if (expired) {
    refreshPromise = (async () => {
      const request = new Request(`${process.env.REACT_APP_API_URL}/auth/refresh`, {
        method: 'POST',
        body: JSON.stringify({accessToken, refreshToken}),
        headers: new Headers({'Content-Type': 'application/json'}),
      });
      const response = await fetch(request);
      if (response.status < 200 || response.status >= 300) {
        throw new HttpError(response.statusText, response.status);
      }
      const data = await response.json();
      localStorage.setItem('authentication.access_token', data.token);
      localStorage.setItem('authentication.refresh_token', data.refreshToken);
    })().finally(() => {
      refreshPromise = null;
    });
    return refreshPromise;
  }
}
```
This will work. However, we're going to provide a way to do that automatically in a later minor version.

**Comment by davidhenley** [comment]
@djhi That was it! Thank you

**Pull request merged**